sim/src/kernels/common.opencl

```diff
@@ -0,0 +1,304 @@
+#ifndef COMMON_OPENCL
```

Comment:
It will become a problem to keep the python prototype/c++ production code coherent with this kind of copying of code. I suggest that we use the Python code as the basis, and perhaps move some of the python notebooks to make it more natural.